accepted/future-releases/abstract-external-fields/feature-specification.md

```diff
@@ -0,0 +1,205 @@
+setter and/or getter which amounts to an implementation, so fields seem to
+be inherently non-abstract and non-external. However, we can identify the
+field with its accessors:
+
+1. An "abstract instance field" can be a pair of an abstract instance
+   getter and an abstract instance setter.  This can be useful, because one
+   declaration replaces two declarations containing some redundant
+   elements: `int get foo;s et foo(int _);`.  Developers could otherwise
+   use a concrete field, `int foo;`, but this may cause a waste of space if
+   someone _extends_ the class instead of implementing it, because the
+   intention is to override the getter and setter, and never use the
+   storage.
+
+2. Similarly, an "external field" can be a pair of an external getter and
+   an external setter.  This came up in connection with `dart:ffi`, where
+   users write Dart classes in order to specify a native memory layout.
+   All instances of the interface will be backed by native (external)
+   code. Currently, `dart:ffi` developers use a regular field declaration,
+   `int foo;`, but what is actually needed is an external setter and/or
+   getter.
+
+An abstract or external field declaration replaces a setter/getter pair or
+a getter.  In the former case it helps avoid redundant specification of the
+name and type, which is more concise and less error-prone.  In the latter
+case the declaration may actually be more verbose (`abstract final int foo;`
+vs. `int get foo;`), but it does no harm to include support for these
+forms, and they may be used to maintain a specific coding style, or it may
+be used by the implementation specific mechanism which is used with a
+certain external field declaration,
+
+A concrete field declaration can be made to work for both situations until
+the null-safety features are introduced.  It is suboptimal, but it works.
+With `dart:ffi`, domain specific rules are used to determine which fields
+are "external", and how they should be connected with the relevant external
+entities.
+
+With null safety that approach no longer works.  For a field whose type
+annotation is a non-nullable type, e.g., `int foo;`, it is an error unless
+initialization is performed by every generative constructor, or the field
+itself is modified to have an initializing expression, e.g.,
+`int foo = 0;`.  This adds yet another element to the declaration which is
+meaningless, confusing for readers of the code, and which may be
+inconvenient to write.  It may even involve nonsensical code like
+`X foo = throw 0;` (where `X` is a type variable) in the case where no
+expression is statically known to evaluate to an instance with the required
+type.
+
+To provide a migration path for these use cases, we introduce _abstract and
+external field declarations_.
+
+
+## Feature Specification
+
+This section specifies abstract and external variables.
+
+
+### Syntax
+
+The grammar is modified as follows:
```

Comment:
Need to include `late` as well. This is happening post-NNBD.

Reply:
It doesn't make sense to include `late` in a grammar rule for a construct that cannot contain that modifier. Or do you wish to include `late abstract` or `late external` variables after all?

Reply:
ACK, it won't apply to the new alternative because it's mutually exclusive with both `abstract` and `external`.